In [43]:
import pandas as pd
import numpy as np
from geopy.geocoders import Nominatim 
import datetime
from sklearn import preprocessing

In [220]:
poke=pd.read_csv("/Users/Shuxian/Documents/CSYE7200-Fall2016/predictemall/300k.csv",low_memory=False,header=0)

#### Drop pokemonId, _id, weatherIcon and S2 columns

poke.columns.get_loc('weatherIcon')

In [221]:
poke.drop(poke.columns[[0,3,4,5,6,7,8,9,10,11,28]],axis=1,inplace=True)

In [222]:
poke.head(5)

latitude   longitude appearedTimeOfDay  appearedHour  appearedMinute  \
0  20.525745  -97.460829             night             5              57   
1  20.523695  -97.461167             night             5              57   
2  38.903590  -77.199780             night             5              57   
3  47.665903 -122.312561             night             5              56   
4  47.666454 -122.311628             night             5              56   

  appearedDayOfWeek  appearedDay  appearedMonth  appearedYear  terrainType  \
0         dummy_day            8              8          2016           14   
1         dummy_day            8              8          2016           14   
2         dummy_day            8              8          2016           13   
3         dummy_day            8              8          2016            0   
4         dummy_day            8              8          2016            0   

   ...   cooc_143 cooc_144 cooc_145 cooc_146  cooc_147  cooc_148  cooc_149  \
0  ...      False    False    False    False     False     False     False   
1  ...      False    False    False    False     False     False     False   
2  ...      False    False    False    False     False     False     False   
3  ...      False    False    False    False     False     False     False   
4  ...      False    False    False    False     False     False     False   

   cooc_150  cooc_151  class  
0     False     False     16  
1     False     False    133  
2     False     False     16  
3     False     False     13  
4     False     False    133  

[5 rows x 197 columns]

In [197]:
poke.appearedYear.unique()

array([2016])

In [198]:
poke.appearedMonth.unique()

array([8])

#### Mon-Sun===>1~7

In [223]:
# poke[poke.columns[[8,7,6]]]
poke['DateTime'] = poke[poke.columns[[8,7,6]]].apply(lambda s : datetime.datetime(*s),axis = 1)

In [224]:
poke['appearedDayOfWeek']=poke['DateTime'].dt.dayofweek
poke['appearedDayOfWeek']=poke['appearedDayOfWeek']+1

In [203]:
poke.head(5)

latitude   longitude appearedTimeOfDay  appearedHour  appearedMinute  \
0  20.525745  -97.460829             night             5              57   
1  20.523695  -97.461167             night             5              57   
2  38.903590  -77.199780             night             5              57   
3  47.665903 -122.312561             night             5              56   
4  47.666454 -122.311628             night             5              56   

   appearedDayOfWeek  appearedDay  appearedMonth  appearedYear  terrainType  \
0                  1            8              8          2016           14   
1                  1            8              8          2016           14   
2                  1            8              8          2016           13   
3                  1            8              8          2016            0   
4                  1            8              8          2016            0   

     ...     cooc_144 cooc_145 cooc_146 cooc_147  cooc_148  cooc_149  \
0    ...        False    False    False    False     False     False   
1    ...        False    False    False    False     False     False   
2    ...        False    False    False    False     False     False   
3    ...        False    False    False    False     False     False   
4    ...        False    False    False    False     False     False   

   cooc_150  cooc_151  class   DateTime  
0     False     False     16 2016-08-08  
1     False     False    133 2016-08-08  
2     False     False     16 2016-08-08  
3     False     False     13 2016-08-08  
4     False     False    133 2016-08-08  

[5 rows x 198 columns]

In [204]:
poke.appearedTimeOfDay.unique()

array(['night', 'evening', 'afternoon', 'morning'], dtype=object)

#### ['afternoon','evening','morning', 'night']==>[1,2,3,4]

In [225]:
le = preprocessing.LabelEncoder()
le.fit(poke.appearedTimeOfDay.unique())
list(le.classes_)
poke['appearedTimeOfDay']=le.transform(poke['appearedTimeOfDay'])+1

In [206]:
poke.head(5)

latitude   longitude  appearedTimeOfDay  appearedHour  appearedMinute  \
0  20.525745  -97.460829                  4             5              57   
1  20.523695  -97.461167                  4             5              57   
2  38.903590  -77.199780                  4             5              57   
3  47.665903 -122.312561                  4             5              56   
4  47.666454 -122.311628                  4             5              56   

   appearedDayOfWeek  appearedDay  appearedMonth  appearedYear  terrainType  \
0                  1            8              8          2016           14   
1                  1            8              8          2016           14   
2                  1            8              8          2016           13   
3                  1            8              8          2016            0   
4                  1            8              8          2016            0   

     ...     cooc_144 cooc_145 cooc_146 cooc_147  cooc_148  cooc_149  \
0    ...        False    False    False    False     False     False   
1    ...        False    False    False    False     False     False   
2    ...        False    False    False    False     False     False   
3    ...        False    False    False    False     False     False   
4    ...        False    False    False    False     False     False   

   cooc_150  cooc_151  class   DateTime  
0     False     False     16 2016-08-08  
1     False     False    133 2016-08-08  
2     False     False     16 2016-08-08  
3     False     False     13 2016-08-08  
4     False     False    133 2016-08-08  

[5 rows x 198 columns]

#### boolean to int

In [226]:
bool_cols=poke.select_dtypes(['bool']).columns

In [227]:
for col in bool_cols:
    poke[col]=poke[col].astype(int)

In [209]:
poke[bool_cols].head(5)

closeToWater  urban  suburban  midurban  rural  gymIn100m  gymIn250m  \
0             0      1         1         1      0          1          1   
1             0      1         1         1      0          0          0   
2             0      0         1         1      0          0          0   
3             1      1         1         1      0          0          0   
4             1      1         1         1      0          0          1   

   gymIn500m  gymIn1000m  gymIn2500m    ...     cooc_142  cooc_143  cooc_144  \
0          1           1           1    ...            0         0         0   
1          1           1           1    ...            0         0         0   
2          1           1           1    ...            0         0         0   
3          1           1           1    ...            0         0         0   
4          1           1           1    ...            0         0         0   

   cooc_145  cooc_146  cooc_147  cooc_148  cooc_149  cooc_150  cooc_151  
0         0         0         0         0         0         0         0  
1         0         0         0         0         0         0         0  
2         0         0         0         0         0         0         0  
3         0         0         0         0         0         0         0  
4         0         0         0         0         0         0         0  

[5 rows x 168 columns]

In [228]:
poke.drop('DateTime',axis=1,inplace=True)
poke.drop('appearedMonth',axis=1,inplace=True)
poke.drop('appearedYear',axis=1,inplace=True)

In [211]:
poke.head(5)

latitude   longitude  appearedTimeOfDay  appearedHour  appearedMinute  \
0  20.525745  -97.460829                  4             5              57   
1  20.523695  -97.461167                  4             5              57   
2  38.903590  -77.199780                  4             5              57   
3  47.665903 -122.312561                  4             5              56   
4  47.666454 -122.311628                  4             5              56   

   appearedDayOfWeek  appearedDay  terrainType  closeToWater         city  \
0                  1            8           14             0  Mexico_City   
1                  1            8           14             0  Mexico_City   
2                  1            8           13             0     New_York   
3                  1            8            0             1  Los_Angeles   
4                  1            8            0             1  Los_Angeles   

   ...   cooc_143 cooc_144  cooc_145  cooc_146  cooc_147  cooc_148  cooc_149  \
0  ...          0        0         0         0         0         0         0   
1  ...          0        0         0         0         0         0         0   
2  ...          0        0         0         0         0         0         0   
3  ...          0        0         0         0         0         0         0   
4  ...          0        0         0         0         0         0         0   

   cooc_150  cooc_151  class  
0         0         0     16  
1         0         0    133  
2         0         0     16  
3         0         0     13  
4         0         0    133  

[5 rows x 195 columns]

#### weather, city, continent to int

In [236]:
le.fit(poke.weather.unique())
list(le.classes_)
poke['weather']=le.transform(poke['weather'])+1

#### global

In [237]:
le.fit(poke.continent.unique())
list(le.classes_)
poke['continent']=le.transform(poke['continent'])+1

In [238]:
le.fit(poke.city.unique())
list(le.classes_)
poke['city']=le.transform(poke['city'])+1

In [ ]:
cols = poke_us.columns.tolist()
cols = cols[-1:] + cols[:-1]
poke_us = poke_us[cols]

In [215]:
poke.to_csv("/Users/Shuxian/Documents/CSYE7200-Fall2016/predictemall/poke.csv",index=False)

#### Filter: return America only, around 150k rows * 200 columns

In [229]:
poke_america=poke.loc[poke['continent'] == 'America']

#### check distinct city names

In [230]:
cities=poke_america.city.unique()

In [231]:
for city in cities:
    city=city.strip()
    city=city.replace("_"," ")
    print(city)

Mexico City
New York
Los Angeles
Chicago
Edmonton
Phoenix
Vancouver
Denver
Toronto
Monterrey
Winnipeg
Montreal
Puerto Rico
Sao Paulo
Halifax
Regina
Guayaquil
Detroit
Bahia
Boise
Bogota
Santiago
Costa Rica
Fortaleza
Belem
Guyana


#### 'New_York', 'Los_Angeles', 'Chicago','Phoenix','Denver','Puerto_Rico','Detroit','Boise

#### US cities only ~120k rows

In [232]:
uscities=['New_York', 'Los_Angeles', 'Chicago','Phoenix','Denver','Puerto_Rico','Detroit','Boise']

In [233]:
poke_us=poke_america[poke_america['city'].isin(uscities)]

In [234]:
poke_us.drop('continent',axis=1,inplace=True)

/Users/Shuxian/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [235]:
le.fit(poke_us.city.unique())
list(le.classes_)
poke_us['city']=le.transform(poke_us['city'])+1

/Users/Shuxian/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()


In [243]:
le.fit(poke_us.weather.unique())
list(le.classes_)
poke_us['weather']=le.transform(poke_us['weather'])+1

/Users/Shuxian/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()


In [246]:
cols = poke_us.columns.tolist()
cols = cols[-1:] + cols[:-1]
poke_us = poke_us[cols]

In [278]:
poke_us.to_csv("/Users/Shuxian/Documents/CSYE7200-Fall2016/predictemall/poke_us.csv",index=False)

geolocator = Nominatim()
location = geolocator.reverse("20.525745, -97.460829")
location.address
location.raw

int(location.raw.get('address').get('postcode'))

poke['coordinate'] = poke[['latitude', 'longitude']].apply(lambda x: ','.join(x), axis=1)

poke['zipcode']=geolocator.reverse((poke.latitude, poke.longitude)).raw.get('address').get('postcode')